pkg/provision/kubernetes.go

```diff
@@ -0,0 +1,311 @@
+package provision
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/threefoldtech/zos/pkg"
+	"github.com/threefoldtech/zos/pkg/stubs"
+)
+
+// Kubernetes reservation data
+type Kubernetes struct {
+	// Size of the vm, this defines the amount of vCpu, memory, and the disk size
+	Size uint8 `json:"size"`
+
+	// NetworkID of the network namepsace in which to run the VM. The network
+	// must be provisioned previously.
+	NetworkID pkg.NetID `json:"network_id"`
+	// IP of the VM. The IP must be part of the subnet available in the network
+	// resource defined by the networkID on this node
+	IP net.IP `json:"ip"`
+
+	// ClusterSecret is the hex encoded encrypted cluster secret.
+	ClusterSecret string `json:"cluster_secret"`
+	// MasterIPs define the URL's for the kubernetes master nodes. If this
+	// list is empty, this node is considered to be a master node.
+	MasterIPs []net.IP `json:"master_ips"`
+	// SSHKeys is a list of ssh keys to add to the VM. Keys can be either
+	// a full ssh key, or in the form of `github:${username}`. In case of
+	// the later, the VM will retrieve the github keys for this username
+	// when it boots.
+	SSHKeys []string `json:"ssh_keys"`
+
+	PlainClusterSecret string `json:"-"`
+}
+
+// TODO: move flist
+const k3osFlistURL = "https://hub.grid.tf/lee/k3os.flist"
+
+func kubernetesProvision(ctx context.Context, reservation *Reservation) (interface{}, error) {
+	return nil, kubernetesProvisionImpl(ctx, reservation)
+}
+
+func kubernetesProvisionImpl(ctx context.Context, reservation *Reservation) error {
```

Comment:
This function doesn't really bring anything, I would just move the code into `kubernetesProvision` function and delete this one.

Reply:
Yes, but this is a convention I suggest (and use) in the other types of provision to make sure the method always return the correct type. You see, the Provision method signature return (interface{}, error) this can allow the code to return Anything or anytime which can make the method inconsistent. To allow the compiler to check for the return type we move the impl to an 'Impl' method, and then add the single line wrapper that return the `(interface{}, error) required signature

Reply:
Yes I know, but here it doesn't return anything. So there is no point doing it at all.

Reply:
But we going to change that as discussed above.

Reply:
right... Forgot this :-)